pkg/storage/azureblob/azureblob.go

```diff
@@ -0,0 +1,151 @@
+package azureblob
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+	"github.com/gomods/athens/pkg/config"
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/observ"
+)
+
+type client interface {
+	UploadWithContext(ctx context.Context, path, contentType string, content io.Reader) error
+	BlobExists(ctx context.Context, path string) (bool, error)
+	ReadBlob(ctx context.Context, path string) (io.ReadCloser, error)
+	ListBlobs(ctx context.Context, prefix string) ([]string, error)
+	DeleteBlob(ctx context.Context, path string) error
+}
+
+type azureBlobStoreClient struct {
+	containerURL *azblob.ContainerURL
+}
+
+func newBlobStoreClient(accountURL *url.URL, accountName, accountKey, containerName string) (*azureBlobStoreClient, error) {
+	cred, err := azblob.NewSharedKeyCredential(accountName, accountKey)
+	if err != nil {
+		return nil, err
+	}
+	pipe := azblob.NewPipeline(cred, azblob.PipelineOptions{})
+	serviceURL := azblob.NewServiceURL(*accountURL, pipe)
+	// rules on container names:
+	// https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#container-names
+	//
+	// This container must exist
+	containerURL := serviceURL.NewContainerURL(containerName)
+	cl := &azureBlobStoreClient{containerURL: &containerURL}
+	return cl, nil
+}
+
+// Storage implements (github.com/gomods/athens/pkg/storage).Saver and
+// also provides a function to fetch the location of a module
+type Storage struct {
+	cl   client
+	conf *config.AzureBlobConfig
+}
+
+// New creates a new azure blobs storage saver
+func New(conf *config.AzureBlobConfig) (*Storage, error) {
+	const op errors.Op = "azure.New"
+	u, err := url.Parse(fmt.Sprintf("https://%s.blob.core.windows.net", conf.AccountName))
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	cl, err := newBlobStoreClient(u, conf.AccountName, conf.AccountKey, conf.ContainerName)
+	if err != nil {
+		return nil, err
+	}
+	return &Storage{cl: cl, conf: conf}, nil
+}
+
+// BlobExists checks if a particular blob exists in the container
+func (c *azureBlobStoreClient) BlobExists(ctx context.Context, path string) (bool, error) {
+	// TODO: Any better way of doing this ?
+	blobURL := c.containerURL.NewBlockBlobURL(path)
+	_, err := blobURL.GetProperties(ctx, azblob.BlobAccessConditions{})
+	if err != nil {
+		var serr azblob.StorageError
+		var ok bool
+
+		if serr, ok = err.(azblob.StorageError); !ok {
+			return false, fmt.Errorf("Error in casting to azure error type")
+		}
+		if serr.Response().StatusCode == http.StatusNotFound {
+			return false, nil
+		}
```

Comment:
Sorry for being nit-picky but what do you think of this?
```
serr, ok = err.(azblob.StorageError)
if !ok {
    return false, fmt.Errorf("Error in casting to azblob.StorageError type")
}
if serr.Response().StatusCode == http.StatusNotFound {
    return false, nil
}
```

Reply:
I'm sorry I didn't get you, please feel free to help me understand the issue. I don't consider myself a go ninja, so any help to improve the code is really appreciated appreciated 🙏

Reply:
It's not really an issue, your code will work, just thought it's more readable/less code that way.